### PR TITLE
Improve Docker images

### LINKED ISSUE #4611: Add Alpine-based Docker images
Couldn't find an issue specifically about Alpine-based versions for the official Docker image, but let me know if one already exists.

### LINKED ISSUE #5269: Docker image consists of two large identical layers for `bun` and `bunx`
### What version of Bun is running?

1.0.1+31aec4ebe325982fc0ef27498984b0ad9969162b

### What platform is your computer?

Darwin 22.6.0 arm64 arm

### What steps can reproduce the bug?

```
docker pull oven/bun:1.0.1

# For detailed analysis:
mkdir bun && cd bun
docker save oven/bun | tar -xz

# Then have a look at the largest files: ./blobs/sha256/{f96a...,7ce0..., 4704...} (all are gzipped tarballs)
```

### What is the expected behavior?

The image is approximately 65 MB large and contains only two larger layers (and some more smaller ones): one is the debian-slim base image (around 30 MB), the other contains the `bun` binary (30–35 MB based on architecture).

### What do you see instead?

The image is 95–100 MB large (depending on architecture).

[Docker Hub](https://hub.docker.com/r/oven/bun/tags):

| DIGEST | OS/ARCH | COMPRESSED SIZE |
| - | - | - |
| 456d511a28e6 | linux/amd64 | 100.52 MB |
| 7190627f958b | linux/arm64 | 94.72 MB |

[Docker Hub Image Details](https://hub.docker.com/layers/oven/bun/latest/images/sha256-456d511a28e648b028180b5c3b0f21b878626c0520647611c543ed075a974b89?context=explore):

| Step | Command | Size |
| - | - | - |
| 1 | `ADD file ... in /` | 29.96 MB |
| 2 | `CMD ["bash"]` | 0 B |
| 3 | `RUN /bin/sh -c groupadd bun` | 3.98 KB |
| 4 | `COPY docker-entrypoint.sh /usr/local/bin # buildkit` | 294 B |
| 5 | `COPY /usr/local/bin/bun /usr/local/bin # buildkit` | 35.28 MB |
| 6 | `COPY /usr/local/bin/bunx /usr/local/bin # buildkit` | 35.28 MB |
| 7 | `WORKDIR /home/bun/app` | 140 B |
| 8 | `ENTRYPOINT ["/usr/local/bin/docker-entrypoint.sh"]` | 0 B |
| 9 | `CMD ["/usr/local/bin/bun"]` | 0 B |

Output during pull:

```
Using default tag: latest
bdf51429e651: Download complete 
7190627f958b: Download complete 
056088840a56: Download complete 
4704d387fc37: Downloading [=========>                                         ]  6.291MB/34.63MB # This is /usr/local/bin/bunx
353472a8cc3f: Download complete 
c297c40960e3: Download complete 
f96ab1515704: Downloading [===========================================>       ]  26.21MB/30.06MB # This is debian-slim
7ce0fd8317b6: Downloading [=========>                                         ]  6.291MB/34.63MB # This is /usr/local/bin/bun
a99d480c3b22: Download complete 
```

### Additional information

In https://github.com/oven-sh/bun/blob/c3455c0ceee6bbe399781819a42fff6cf24792e2/dockerhub/Dockerfile-debian#L56 we see that `bunx` is a symbolic link pointing to `bun`. However, symbolic links are resolved by the Dockerfile `COPY` command (mentioned e.g. in https://github.com/moby/moby/issues/40449), so the resulting image contains the exact same file twice, in two different layers.

### PR DESCRIPTION
**Pull request by Electroid**
Closes #5269 
Closes #4611 
Closes #5454

### CODE REVIEW

File: dockerhub/debian-slim/Dockerfile [hunk 1]
```diff
@@ -69,9 +66,8 @@ RUN groupadd bun \
       --create-home
 
 COPY docker-entrypoint.sh /usr/local/bin
-COPY --from=build /usr/local/bin/bun /usr/local/bin
-COPY --from=build /usr/local/bin/bunx /usr/local/bin
+COPY --from=build /usr/local/bin/bun /usr/local/bin/bun
+RUN ln -s /usr/local/bin/bun /usr/local/bin/bunx
```
Comment (Jarred-Sumner) [inline dockerhub/debian-slim/Dockerfile]: should it be a hardlink instead? so that it's not dependent on filesystem paths

**Comment by Electroid** [comment]
Need to update GitHub actions, after that will be good to go.

**Comment by jumoog** [comment]
You should add `--retry-delay` ( for example `--retry-delay 30` )  to curl. Otherwise, curl will use an exponential backoff algorithm.

> When curl is about to retry a transfer, it will first wait one second and then for all forthcoming retries it will double the waiting time until it reaches 10 minutes which then will be the delay between the rest of the retries. Using `--retry-delay` you can disable this exponential backoff algorithm and set your own delay between the attempts

Source: https://everything.curl.dev/usingcurl/downloads/retry#tweak-your-retries


This means:

| request | delay in seconds |
|---------|------------------|
| 1       | 1                |
| 2       | 2                |
| 3       | 4                |
| 4       | 8                |
| 5       | 16               |

Using the exponential backoff algorithm results in an overall delay of only 31 seconds.

File: dockerhub/distroless/Dockerfile [hunk 2]
```diff
@@ -34,36 +33,35 @@ RUN apt-get update -qq \
...
 FROM gcr.io/distroless/base-nossl-debian11
 
-COPY --from=build /usr/local/bin/bun /usr/local/bin
-COPY --from=build /usr/local/bin/bunx /usr/local/bin
+COPY --from=build /usr/local/bin/bun /usr/local/bin/
+
+# Known issue: `bunx` is not available in distroless.
+#
+# If `ln` is used in the build image, the size of the final
+# image will be double, because of: https://github.com/oven-sh/bun/issues/5269
```
Comment (polarathene) [inline dockerhub/distroless/Dockerfile:65]: **UPDATE:** I opened a PR: https://github.com/oven-sh/bun/pull/6090

---

You missed my comment?: https://github.com/oven-sh/bun/issues/5269#issuecomment-1724856820

You can keep it:

```Dockerfile
# List of sources to destination (final path):
COPY --from=build \
  /usr/local/bin/bun /usr/local/bin/bunx \
  /usr/local/bin
```

Comparision:

```console
# Resolved with single COPY, symlink works:
$ docker save local-bun-build | gzip -c | wc -c | numfmt --to iec
40M

# The duplicate issue that previously existed:
$ docker save oven/bun | gzip -c | wc -c | numfmt --to iec
97M

# Resolved vs original image size:
$ docker images
REPOSITORY                  TAG                IMAGE ID       CREATED          SIZE
local-bun-build             latest             020acf83835e   7 minutes ago    110MB
oven/bun                    latest             647ebb4444b7   3 days ago       271MB
```

**Pull request merged**